granulate_utils/metrics/sampler.py

```diff
@@ -0,0 +1,240 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+import os
+import socket
+from threading import Event, Thread
+from typing import Any, Optional, Tuple
+from xml.etree import ElementTree as ET
+
+import psutil
+
+from granulate_utils.exceptions import MissingExePath
+from granulate_utils.linux.ns import resolve_host_path
+from granulate_utils.linux.process import process_exe, search_for_process
+from granulate_utils.metrics.mode import SPARK_MESOS_MODE, SPARK_STANDALONE_MODE, SPARK_YARN_MODE
+
+FIND_CLUSTER_TIMEOUT_SECS = 10 * 60
+
+
+class SparkSampler:
+    """
+    Spark cluster metrics sampler
+    """
+
+    def __init__(
+        self,
+        sample_period: float,
+        storage_dir: str,
+        logger: Any,
+    ):
+        self._logger = logger
+        self._master_address: Optional[str] = None
+        self._spark_mode: Optional[str] = None
+        self._collection_thread: Optional[Thread] = None
+        self._sample_period = sample_period
+        # not the same instance as GProfiler._stop_event
+        self._stop_event = Event()
+        self._stop_collection = False
+        self._is_running = False
+        self._storage_dir = storage_dir
+        if self._storage_dir is not None:
+            assert os.path.exists(self._storage_dir) and os.path.isdir(self._storage_dir)
+        else:
+            self._logger.debug("Output directory is None. Will add metrics to queue")
+
+    def get_hostname(self) -> str:
+        try:
+            hostname = socket.gethostname()
+        except Exception as e:
+            self._logger.exception("Failed to get hostname", exception=e)
+        return hostname
```

Comment:
This is incorrect to get hostname that way. gProfiler gets the hostname in the outer network & mount namespace.

Let's keep this code outside of this class - it should accept hostname as a parameter, then gProfiler can pass the hostname it has.

Generally when moving code - DO NOT resort to re-implementing things because it's inconvenient to move them. Let's discuss each particular case.